charmcraft/config.py

```diff
@@ -0,0 +1,54 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+"""Central configuration management."""
+
+import pathlib
+
+from charmcraft.cmdbase import CommandError
+from charmcraft.utils import load_yaml
+
+from collections import UserDict
+
+
+class _Config(UserDict):
+    """Hold the config from charmcraft.yaml."""
+
+    def __init__(self):
+        self.data = {}
```

Comment:
I've been pushing Snapcraft away from (ab)using dictionaries and instead making this type of object into first-class python object to make it easier to use/maintain. Something like:

```
class Config:
  def __init__(self, *, name: str) -> None:
    self.name = name

  def validate(self) -> None:
    """Validate all properties."""
    ...

  @classmethod
  def unmarshal(self, raw_data) -> "Config":
    """Validate the schema & mypy typing assumptions and return Config."""
    if not isinstance(raw_data, dict):
            raise CommandError("Invalid charmcraft.yaml structure: must be a dictionary.")
    name = raw_data.get("name")
    if not isinstance(name, str):
            raise CommandError(f"Charm name {name!r} is not a valid string.")
    config = Config(name=name)
    config.validate()
    return config
````

Reply:
As we discussed, I'll go to a schema oriented complex object for this config.